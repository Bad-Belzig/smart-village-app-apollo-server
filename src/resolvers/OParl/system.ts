import { FilterQuery } from 'mongoose';

import { Body, ISystem, ISystemSchema, System } from '../../models';
import { getPaginatedEntriesByIds } from '../resolverHelpers';

export const systemResolvers = {
  Query: {
    oParlSystems: (_, args: { externalIds?: string[]; keyword?: string[] }) => {
      const filter: FilterQuery<ISystemSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };

      return System.find(filter);
    },
  },
  OParlSystem: {
    body: async (
      parent: ISystem,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(Body, parent.body, args.offset, args.pageSize),
  },
};
