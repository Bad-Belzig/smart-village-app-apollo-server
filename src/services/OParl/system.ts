import { System } from '../../models';
import { parseSystem } from '../../parser';
import { getJson, updateOrCreateEntry } from './importHelpers';
import { ImportQueue, ImportQueueEntry, ImportType } from './ImportTypes';
import { fetchPaginatedOParlList } from './paginatedList';

// import flow
// 1. fetch json if it was not already the whole json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importSystem = async (value: unknown, queue?: ImportQueue) => {
  const json = await getJson(value);

  if (!json) return;

  const addToQueue: [ImportQueueEntry | ImportQueueEntry[], ImportType][] = [];

  if (json.body) {
    json.body = await fetchPaginatedOParlList(json.body);
    addToQueue.push([json.body, ImportType.Body]);
  }

  try {
    return await updateOrCreateEntry(
      json,
      parseSystem,
      System,
      addToQueue,
      queue,
    );
  } catch (e) {
    console.log(`Error while importing system from ${value}:`, e);
  }
};
