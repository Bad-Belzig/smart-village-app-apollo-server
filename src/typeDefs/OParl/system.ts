import { gql } from 'apollo-server-core';
import { oParlBase } from './oParlBase';

export const systemGQL = gql`
  extend type Query {
    oParlSystems(
      externalIds: [String!]
      keyword: [String!]
      offset: Int
      pageSize: Int
    ): [OParlSystem!]
  }

  type OParlSystem {
    body(offset: Int, pageSize: Int): [OParlBody!]
    oparlVersion: String!
    contactEmail: String
    contactName: String
    name: String
    otherOparlVersions: [String!]
    product: String
    vendor: String
    website: String
    ${oParlBase}
  }
`;
