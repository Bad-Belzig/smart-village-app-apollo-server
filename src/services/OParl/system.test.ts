import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../testHelpers';
import { importSystem } from './system';

const helperFuncs = setupTestDataBase('test_system_import');

// TODO: change test url to something within our control
const testUrl = 'https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/system.asp';

describe('importing a system', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importSystem, testUrl));

  it('import returns undefined on faulty url', invalidUrlTest(importSystem));

  afterAll(helperFuncs.afterAll);
});
