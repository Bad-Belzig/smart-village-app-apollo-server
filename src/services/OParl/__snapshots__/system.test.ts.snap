// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a system imports properly 1`] = `
Object {
  "__v": 0,
  "body": Array [],
  "contactEmail": "andreas.schreiber@kreis-lup.de",
  "contactName": "Andreas Schreiber",
  "externalId": "https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/system.asp",
  "keyword": Array [],
  "license": "http://opendatahandbook.org/guide/de/how-to-open-up-data/",
  "name": "Landkreis LUP",
  "oparlVersion": "https://schema.oparl.org/1.0/",
  "otherOparlVersions": Array [],
  "product": "http://cc-egov.de/produkte/allris",
  "type": "https://schema.oparl.org/1.0/System",
  "vendor": "http://cc-egov.de/",
  "website": "https://www.lwl-pch.sitzung-online.de/bi/",
}
`;
